### ISSUE TITLE
HeatMap and PolarHeatMap interpolation for non-uniform grids

### ISSUE THREAD

**Issue by pbandyk** [oxyplot/oxyplot#911]
### Feature description

The current implementation in HeatMapSeries assumes regularly spaced grid, using only the lower and upper ends of the axis being used and the number of data points.  For example, we set X0=0 and X1=10 and have 11 data points along that dimension, giving an assumed spacing of 1.  It would be nice if the interpolation algorithm could handle irregular spacing, as done by ContourSeries.

The same comment applies to PolarHeatMapSeries.

**Comment by objorke** [comment]
HeatMapSeries: it should be possible to do this when rendering by rectangles (not bitmap). I think it already supports log axes!

**Comment by dynamatt** [comment]
I'm going to try and implement this by adding a mapping property similar to the DataPointSeries.

**Comment by dynamatt** [comment]
Hi guys, I have a working implementation of a HeatMapSeries which can be mapped to an ItemsSource. I think it's a bit rough though. A working example can be found in the "HeatMapMappingDemo" in the WPF Examples.

I was wondering if somebody could code-review what I've done so far and provide some advice?

**Comment by objorke** [comment]
https://github.com/oxyplot/oxyplot/compare/develop...dynamatt:%23911
I will have a look later!

**Comment by objorke** [comment]
let us handle the polar case in a separate issue - make it easier to close this issue

**Comment by objorke** [comment]
Could this be implemented by creating a coordinate definition classes:
```csharp
enum CoordinatePosition { Center, Edge }
class CoordinateDefinition {
    CoordinatePosition Position
}
class UniformLinearCoordinateDefinition : CoordinateDefinition { 
    double X0
    double X1
}
class CustomCoordinateDefinition {
  double[] Coordinates
}
```
then in the series: 
```csharp
class HeatMapSeries {
  CoordinateDefinition RowCoordinateDefinition
  CoordinateDefinition ColumnCoordinateDefinition
}
class ContourSeries {
  CoordinateDefinition RowCoordinateDefinition
  CoordinateDefinition ColumnCoordinateDefinition
}
```
and get rid of X0,X1, Y0, Y1 (HeatMapSeries) and  RowCoordinates and ColumnCoordinates (ContourSeries)

**Comment by greg-stein** [comment]
Hi guys,

Sorry for necroposting, but I couldn't find a working example of the non uniform grid for HeatMapSeries. Is it supported?

Kindest regards, Greg.

**Comment by VisualMelon** [comment]
No, this isn't included at present. A proposal like objorke's above would probably be a good idea.

**Comment by 23W** [comment]
Hi guys,
Has irregular spaced interpolation been implemented in `HeatMapSeries` or not?

**Comment by VisualMelon** [comment]
No: the implementation does not support non-even spacing. Would need an overhaul.
